{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e72bda4e-4c09-4d45-9026-a056d3671f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "# Change the working directory\n",
    "os.chdir('/private/home/yxu267/anaconda3/envs/dimelo/lib/python3.10/site-packages')\n",
    "\n",
    "# Verify the change\n",
    "print(os.getcwd())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5a3b72b-4f9f-4f4f-aebe-5ae2015ddb9c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98e52368",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import ast\n",
    "\n",
    "\n",
    "\n",
    "column_names = ['Chromosome', 'Start/End', 'Density', 'Coverage']\n",
    "# Load the CSV file into a DataFrame\n",
    "CENPA_chrom_arm_df = pd.read_csv('CENPA_chromosome_arm_random_region_dict_region_density_scores.csv',\n",
    "                              header=None,\n",
    "                              names=column_names,\n",
    "                              sep='\\t')\n",
    "\n",
    "CENPA_CDR_df = pd.read_csv('/private/groups/migalab/dan/data_analysis/mC_border_analysis/CENPA_CDR_dict_region_density_scores.csv',\n",
    "                           header=None,\n",
    "                           names=column_names,\n",
    "                              sep='\\t')\n",
    "\n",
    "CENPA_non_active_df = pd.read_csv('/private/groups/migalab/dan/data_analysis/mC_border_analysis/none_CDR_active_dict_region_density_scores.csv',\n",
    "                                  header=None,\n",
    "                                  names=column_names,\n",
    "                              sep='\\t')\n",
    "\n",
    "#Thes lines remove the X and Y chromosomes \n",
    "\n",
    "CENPA_chrom_arm_df = CENPA_chrom_arm_df[~CENPA_chrom_arm_df['Chromosome'].str.contains('chrX_MATERNAL')]\n",
    "CENPA_chrom_arm_df = CENPA_chrom_arm_df[~CENPA_chrom_arm_df['Chromosome'].str.contains('chrY_PATERNAL')]\n",
    "\n",
    "CENPA_non_active_df = CENPA_non_active_df[~CENPA_non_active_df['Chromosome'].str.contains('chrX_MATERNAL')]\n",
    "CENPA_non_active_df = CENPA_non_active_df[~CENPA_non_active_df['Chromosome'].str.contains('chrY_PATERNAL')]\n",
    "\n",
    "CENPA_CDR_df = CENPA_CDR_df[~CENPA_CDR_df['Chromosome'].str.contains('chrX_MATERNAL')]\n",
    "CENPA_CDR_df = CENPA_CDR_df[~CENPA_CDR_df['Chromosome'].str.contains('chrY_PATERNAL')]\n",
    "\n",
    "chromosomes = [f'chr{i}' for i in range(1, 23)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76501d0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "#parse CDR based on A count \n",
    "\n",
    "\n",
    "assembly_ = open(\"/private/groups/migalab/dan/reference/hg002v1.0.1.fasta\", \"r\")\n",
    "\n",
    "#Load the reference genome and make it into a dictionary \n",
    "fasta_sequences = SeqIO.parse(assembly_, \"fasta\")\n",
    "assembly={}\n",
    "for fasta in fasta_sequences:\n",
    "    name, sequence = fasta.id, str(fasta.seq)\n",
    "    assembly[name] = sequence\n",
    "\n",
    "#Make a dictionary for all the chromosomes and their corresponding sequence length \n",
    "assembly_sequence_length = {}    \n",
    "for chromosome in assembly:\n",
    "    assembly_sequence_length[chromosome] = len(assembly[chromosome])\n",
    "    \n",
    "assembly_.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a7feff4",
   "metadata": {},
   "outputs": [],
   "source": [
    "   \n",
    "def comprehensive_mA_density_calculation(region_df, assembly):   \n",
    "    region_weighted_df = pd.DataFrame(columns=['Chromosome', 'Difference', 'Average'])\n",
    "    region_combined_df = pd.DataFrame(columns=['Chromosome', 'Start/End', 'Density'])\n",
    "\n",
    "    # Filter rows for chromosome 1 to 22\n",
    "    chromosomes = [f'chr{i}' for i in range(1, 23)]\n",
    "\n",
    "    for chrom in chromosomes:\n",
    "\n",
    "        maternal_col = f'{chrom}_MATERNAL'\n",
    "        paternal_col = f'{chrom}_PATERNAL'\n",
    "\n",
    "        #grab chromosome hap coordinates and put them in seperate lists \n",
    "        MAT_indices = region_df.index[region_df['Chromosome'] == maternal_col].tolist()\n",
    "        PAT_indices = region_df.index[region_df['Chromosome'] == paternal_col].tolist()\n",
    "\n",
    "        #make the start and end positions of each and every chromosome into and individual list \n",
    "        PAT_CDR_coordinates_from_density = region_df.loc[PAT_indices]['Start/End'].tolist()\n",
    "        MAT_CDR_coordinates_from_density = region_df.loc[MAT_indices]['Start/End'].tolist()\n",
    "        \n",
    "        PAT_CDR_coordinates_total_A = 0 \n",
    "        PAT_CDR_coordinates_total_mA = 0 \n",
    "\n",
    "        # calculate the total amount of mA from the total amount of As in the reference genome and density \n",
    "        # calculated in the previous script in both Paternal and Maternal chromosomes \n",
    "        for coordinate in range (len(PAT_CDR_coordinates_from_density)): \n",
    "            PAT_CDR_coordinates_start_end = ast.literal_eval(PAT_CDR_coordinates_from_density[coordinate])\n",
    "            PAT_CDR_coordinates_start = int (PAT_CDR_coordinates_start_end[0])\n",
    "            PAT_CDR_coordinates_end = int (PAT_CDR_coordinates_start_end[1])\n",
    "            PAT_CDR_A_count = assembly[paternal_col][PAT_CDR_coordinates_start:PAT_CDR_coordinates_end].count('A')\n",
    "            PAT_CDR_coordinates_total_A += PAT_CDR_A_count\n",
    "            PAT_CDR_coordinates_total_mA += PAT_CDR_A_count * region_df['Density'][PAT_indices[coordinate]]\n",
    "\n",
    "\n",
    "        PAT_CDR_coordinates_comprehensive_density = PAT_CDR_coordinates_total_mA / PAT_CDR_coordinates_total_A\n",
    "        new_row = pd.DataFrame({\n",
    "                                    'Chromosome': [paternal_col],\n",
    "                                    'Start/End': [[PAT_CDR_coordinates_start, PAT_CDR_coordinates_end]],\n",
    "                                    'Density': [PAT_CDR_coordinates_comprehensive_density]\n",
    "                                })\n",
    "\n",
    "        region_combined_df = pd.concat([region_combined_df, new_row], ignore_index=True)\n",
    "\n",
    "\n",
    "        MAT_CDR_coordinates_total_A = 0 \n",
    "        MAT_CDR_coordinates_total_mA = 0 \n",
    "\n",
    "        for coordinate in range (len(MAT_CDR_coordinates_from_density)): \n",
    "            MAT_CDR_coordinates_start_end = ast.literal_eval(MAT_CDR_coordinates_from_density[coordinate])\n",
    "            MAT_CDR_coordinates_start = int (MAT_CDR_coordinates_start_end[0])\n",
    "            MAT_CDR_coordinates_end = int (MAT_CDR_coordinates_start_end[1])\n",
    "            MAT_CDR_A_count = assembly[maternal_col][MAT_CDR_coordinates_start:MAT_CDR_coordinates_end].count('A')\n",
    "            MAT_CDR_coordinates_total_A += MAT_CDR_A_count\n",
    "            MAT_CDR_coordinates_total_mA += MAT_CDR_A_count * region_df['Density'][MAT_indices[coordinate]]\n",
    "   \n",
    "        MAT_CDR_coordinates_comprehensive_density = MAT_CDR_coordinates_total_mA / MAT_CDR_coordinates_total_A\n",
    "        new_row = pd.DataFrame({\n",
    "                                    'Chromosome': [maternal_col],\n",
    "                                    'Start/End': [[MAT_CDR_coordinates_start, MAT_CDR_coordinates_end]],\n",
    "                                    'Density': [MAT_CDR_coordinates_comprehensive_density]\n",
    "                                })\n",
    "        region_combined_df = pd.concat([region_combined_df, new_row], ignore_index=True)\n",
    "\n",
    "\n",
    "        chromosome_average = (PAT_CDR_coordinates_comprehensive_density + MAT_CDR_coordinates_comprehensive_density) / 2\n",
    "\n",
    "\n",
    "        diff = abs (((PAT_CDR_coordinates_comprehensive_density - MAT_CDR_coordinates_comprehensive_density)/ MAT_CDR_coordinates_comprehensive_density)*100)\n",
    "        regional_new_row = pd.DataFrame({\n",
    "                                    'Chromosome': [chrom],\n",
    "                                    'Difference': [diff],\n",
    "                                    'Average': [chromosome_average]\n",
    "                                })\n",
    "        \n",
    "        region_weighted_df = pd.concat([region_weighted_df, regional_new_row], ignore_index=True)\n",
    "    return region_weighted_df, region_combined_df\n",
    "\n",
    "CENPA_CDR_A_weighted_df, CENPA_CDR_A_combined_df = comprehensive_mA_density_calculation(CENPA_CDR_df, assembly)\n",
    "CENPA_chrom_arm_weighted_df, CENPA_chrom_arm_A_combined_df = comprehensive_mA_density_calculation(CENPA_chrom_arm_df, assembly)\n",
    "CENPA_non_active_weighted_df, CENPA_non_active_A_combined_df = comprehensive_mA_density_calculation(CENPA_non_active_df, assembly)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df4c6181",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ba85492",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Function to process the data and create a DataFrame for strip plot\n",
    "def process_df(df, type_label,type_pos):\n",
    "    strip_plot_df = pd.DataFrame(columns=['Chromosome', 'Density', 'Type','Type_Pos'])\n",
    "    new_rows = []\n",
    "    for chrom in chromosomes:\n",
    "        maternal_col = f'{chrom}_MATERNAL'\n",
    "        paternal_col = f'{chrom}_PATERNAL'\n",
    "\n",
    "        MAT_indices = df.index[df['Chromosome'] == maternal_col].tolist()\n",
    "        PAT_indices = df.index[df['Chromosome'] == paternal_col].tolist()\n",
    "\n",
    "        PAT_values = df.loc[PAT_indices]['Density'].tolist()\n",
    "        MAT_values = df.loc[MAT_indices]['Density'].tolist()\n",
    "\n",
    "        PAT_density_avg = np.mean(PAT_values) if PAT_values else 0\n",
    "        MAT_density_avg = np.mean(MAT_values) if MAT_values else 0\n",
    "\n",
    "        # Append new row data for paternal and maternal columns to the list\n",
    "        new_rows.append({'Chromosome': paternal_col,\n",
    "                         'Density': PAT_density_avg,\n",
    "                         'Type': f'PAT_{type_label}',\n",
    "                         'Type_Pos': type_pos})\n",
    "        \n",
    "        new_rows.append({'Chromosome': maternal_col,\n",
    "                         'Density': MAT_density_avg,\n",
    "                         'Type': f'MAT_{type_label}',\n",
    "                         'Type_Pos': type_pos + 0.2})\n",
    "\n",
    "    # Create a DataFrame from the list of new rows and concatenate with the existing DataFrame\n",
    "    new_rows_df = pd.DataFrame(new_rows)\n",
    "    strip_plot_df = pd.concat([strip_plot_df, new_rows_df], ignore_index=True)\n",
    "\n",
    "    return strip_plot_df\n",
    "\n",
    "# Assuming 'chromosomes' and the DataFrames are already defined\n",
    "\n",
    "strip_plot_non_active_df = process_df(CENPA_non_active_A_combined_df, 'non_CDR_active',0.5)\n",
    "strip_plot_CDR_df = process_df(CENPA_CDR_A_combined_df, 'CDR', 1.6)\n",
    "strip_plot_chrom_arm_df = process_df(CENPA_chrom_arm_A_combined_df, 'chrom_arm', 1.0)\n",
    "\n",
    "# Concatenate the processed DataFrames\n",
    "combined_df = pd.concat([strip_plot_non_active_df, \n",
    "                         strip_plot_chrom_arm_df,\n",
    "                         strip_plot_CDR_df], ignore_index=True)\n",
    "\n",
    "# Update the x-coordinates mapping for plotting\n",
    "type_pos_map = {'PAT_non_CDR_active': 0.5,\n",
    "                'MAT_non_CDR_active': 0.7,\n",
    "               'PAT_chrom_arm':1.0,\n",
    "               'MAT_chrom_arm':1.2,\n",
    "                'PAT_CDR': 1.6,\n",
    "                'MAT_CDR': 1.8}\n",
    "#combined_df['Type_Pos'] = combined_df['Type'].map(type_pos_map)\n",
    "print (combined_df)\n",
    "\n",
    "medians_dict = {type_label: combined_df[combined_df['Type'] == type_label]['Density'].median() for type_label in type_pos_map.keys()}\n",
    "\n",
    "# Plotting\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.scatterplot(x='Type_Pos', y='Density', hue='Type', style='Type', data=combined_df, s=100, palette=['blue', 'red', 'green', 'orange','black','pink'])\n",
    "\n",
    "# Drawing lines between the points\n",
    "for chrom in chromosomes:\n",
    "    for type_label in ['non_CDR_active', 'CDR','chrom_arm']:\n",
    "        pat_row = combined_df[(combined_df['Chromosome'] == f'{chrom}_PATERNAL') & (combined_df['Type'] == f'PAT_{type_label}')]\n",
    "        mat_row = combined_df[(combined_df['Chromosome'] == f'{chrom}_MATERNAL') & (combined_df['Type'] == f'MAT_{type_label}')]\n",
    "        if not pat_row.empty and not mat_row.empty:\n",
    "            plt.plot([type_pos_map[f'PAT_{type_label}'], type_pos_map[f'MAT_{type_label}']],\n",
    "                     [pat_row['Density'].values[0], mat_row['Density'].values[0]],\n",
    "                     color='gray', linestyle='--')\n",
    "\n",
    "for type_label, median in medians_dict.items():\n",
    "    plt.bar(type_pos_map[type_label], median, width=0.1, color='blue', alpha=0.5)\n",
    "\n",
    "            \n",
    "# Customizing the plot\n",
    "plt.xticks(list(type_pos_map.values()), list(type_pos_map.keys()))\n",
    "plt.title('CENPA density Values between homologs in adaptive sequencing')\n",
    "plt.xlabel('Type')\n",
    "plt.ylabel('Density')\n",
    "plt.legend(title='Type', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08b8abe0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Assuming 'CENPA_CDR_A_combined_df' is already defined and contains 'Chromosome', 'Density', and we already have 'Chromosome_Label'\n",
    "# Add a new column to indicate if the data is 'Paternal' or 'Maternal'\n",
    "CENPA_CDR_A_combined_df['Chromosome_Label'] = CENPA_CDR_A_combined_df['Chromosome'].str.replace('_PATERNAL', '').str.replace('_MATERNAL', '')\n",
    "\n",
    "CENPA_CDR_A_combined_df['Type'] = CENPA_CDR_A_combined_df['Chromosome'].apply(lambda x: 'Paternal' if 'PATERNAL' in x else 'Maternal')\n",
    "\n",
    "# Set the color palette\n",
    "palette = {'Paternal': 'blue', 'Maternal': 'red'}\n",
    "\n",
    "# Plotting\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.barplot(x='Chromosome_Label', y='Density', hue='Type', data=CENPA_CDR_A_combined_df, palette=palette, errorbar=None)\n",
    "\n",
    "plt.title('CENPA CDR density values between homologs in adaptive sequencing')\n",
    "plt.xlabel('Chromosome')\n",
    "plt.ylabel('Density')\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02bfb550",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
