{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f1055690",
   "metadata": {},
   "source": [
    "### Inputs:\n",
    "- `bamfile`: Path to the BAM file containing alignment data.\n",
    "- `assembly_`: Path to the reference genome in FASTA format.\n",
    "\n",
    "### Outputs:\n",
    "- `assembly`: Dictionary containing chromosome names as keys and their sequences as values.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': 'ATCGATCG...',\n",
    "        'chr2': 'GCTAGCTA...'\n",
    "    }\n",
    "    ```\n",
    "- `assembly_sequence_length`: Dictionary containing chromosome names as keys and their sequence lengths as values.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': 248956422,\n",
    "        'chr2': 242193529\n",
    "    }\n",
    "    ```\n",
    "- Prints the time taken to load the reference genome and create the dictionaries.\n",
    "\n",
    "### Description:\n",
    "This code imports necessary libraries, sets up the environment, and loads alignment data from a BAM file. It reads a reference genome from a FASTA file, creating dictionaries to store chromosome sequences and their lengths. Additionally, it measures and prints the time taken to perform these operations.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b835e59",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import time \n",
    "import matplotlib.patches as patches\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.pyplot as plt\n",
    "import itertools\n",
    "import pandas as pd \n",
    "from tabulate import tabulate\n",
    "import csv\n",
    "import random\n",
    "\n",
    "import os\n",
    "# Change the working directory\n",
    "os.chdir('/private/home/yxu267/anaconda3/envs/dimelo/lib/python3.10/site-packages')\n",
    "\n",
    "# Verify the change\n",
    "print(os.getcwd())\n",
    "\n",
    "\n",
    "\n",
    "np.set_printoptions(threshold=np.inf)\n",
    "min_quality_score = 8\n",
    "\n",
    "#Load the bam file \n",
    "bamfile = pysam.AlignmentFile(\n",
    "    \"/private/groups/migalab/dan/06_11_24_R1041_UL_DiMeLo_CENPAyoung_1/20240611_1126_1H_PAW33460_814408d8/pod5/06_11_24_R1041_UL_DiMeLo_CENPAyoung_1_5mA_6mC_winnowmap_MD_mA_mC.bam\",\n",
    "    \"rb\") \n",
    "\n",
    "\n",
    "assembly_ = open(\"/private/groups/migalab/dan/reference/hg002v1.0.1.fasta\", \"r\")\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "#Load the reference genome and make it into a dictionary \n",
    "fasta_sequences = SeqIO.parse(assembly_, \"fasta\")\n",
    "assembly={}\n",
    "for fasta in fasta_sequences:\n",
    "    name, sequence = fasta.id, str(fasta.seq)\n",
    "    assembly[name] = sequence\n",
    "\n",
    "#Make a dictionary for all the chromosomes and their corresponding sequence length \n",
    "assembly_sequence_length = {}    \n",
    "for chromosome in assembly:\n",
    "    assembly_sequence_length[chromosome] = len(assembly[chromosome])\n",
    "    \n",
    "end_time = time.time()\n",
    "elapsed_time = end_time - start_time\n",
    "print (elapsed_time, \"seconds\")\n",
    "assembly_.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88c807c3",
   "metadata": {},
   "source": [
    "### Inputs:\n",
    "- `input_file`: Path to the BED file containing chromosome and active array data.\n",
    "\n",
    "### Outputs:\n",
    "- `active_dict`: Dictionary where chromosome names are keys and the values are lists of active array blocks.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': [[1000, 2000], [3000, 4000]],\n",
    "        'chr2': [[500, 1500]]\n",
    "    }\n",
    "    ```\n",
    "\n",
    "### Description:\n",
    "The purpose of this code is to create an active array dictionary where each chromosome name serves as a key. The corresponding value is a list of active array blocks represented by their start and end positions. The code reads the input BED file, processes each line to extract chromosome information and active array positions, and populates the `active_dict` dictionary accordingly. It handles changes in chromosome number and parental status to reset or continue the numbering of active blocks as needed.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d685ed4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''The purpose of the code here is to make an active array where the chromosome name is\n",
    "the dictionary key. The active array blocks are in lists inside the key.\n",
    "'''\n",
    "import os\n",
    "input_file = '/private/groups/migalab/dan/data_analysis/alpha_bed/hg002v1.0.fasta.manualAlpha.cenSat_H1L_merged.bed'\n",
    "active_dict = {}\n",
    "with open(input_file, 'r') as infile:  \n",
    "    num = 0 \n",
    "    previous_chr_num = \"\"\n",
    "    previous_parental_status = \"\"\n",
    "    for i in infile:\n",
    "        chr_num = i.split ('_')[0]\n",
    "        parental_status = i.split ('_')[1][0:8]\n",
    "        if chr_num == previous_chr_num and parental_status !=previous_parental_status:\n",
    "            num = 0 \n",
    "        elif chr_num == previous_chr_num and parental_status ==previous_parental_status:\n",
    "            pass \n",
    "        else: \n",
    "            num = 0 \n",
    "        active = i.split ()\n",
    "        if active[0] not in active_dict:\n",
    "            \n",
    "            active_dict[active[0]] = [[int(active[1]) ,int(active[2])]]\n",
    "        else:\n",
    "            active_dict[active[0]].append([int(active[1]) ,int(active[2])])\n",
    "        previous_chr_num = chr_num \n",
    "        previous_parental_status = parental_status\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8975b3da",
   "metadata": {},
   "source": [
    "### Inputs:\n",
    "- `active_dict`: Dictionary where chromosome names are keys and the values are lists of active array blocks.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': [[1000, 2000], [3000, 4000]],\n",
    "        'chr2': [[500, 1500]]\n",
    "    }\n",
    "    ```\n",
    "\n",
    "### Outputs:\n",
    "- `active_array_range`: Dictionary where chromosome names are keys and the values are lists of active array sub-ranges, each sub-range having a length of 1000.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': [[1000, 2000], [2000, 3000], [3000, 4000]],\n",
    "        'chr2': [[500, 1500]]\n",
    "    }\n",
    "    ```\n",
    "\n",
    "### Description:\n",
    "This code creates a new dictionary, `active_array_range`, which contains chromosome names as keys and lists of active array sub-ranges as values. For each chromosome in the input dictionary `active_dict`, the code iterates over the active array blocks and subdivides each block into smaller ranges of 1000 base pairs. These sub-ranges are then added to the `active_array_range` dictionary. The resulting dictionary provides a more granular representation of the active array blocks.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cf984e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "active_array_range={}\n",
    "for chromosome in active_dict:\n",
    "    active_array_range[chromosome] = []\n",
    "    for fragment in range (active_dict[chromosome][0][0], active_dict[chromosome][0][1],1000):\n",
    "\n",
    "        active_array_range[chromosome].append ([fragment,fragment+1000])\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de7ab5b7",
   "metadata": {},
   "source": [
    "### Inputs:\n",
    "- `input_file`: Path to the BED file containing CDR regions data.\n",
    "\n",
    "### Outputs:\n",
    "- `CDR_dict`: Dictionary where chromosome names are keys and the values are lists of CDR regions represented by their start and end positions.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': [[1000, 2000], [3000, 4000]],\n",
    "        'chr2': [[500, 1500]]\n",
    "    }\n",
    "    ```\n",
    "\n",
    "### Description:\n",
    "This code formulates the CDR (Conserved Domain Region) regions and lists them for each chromosome. It reads the input BED file and processes each line to extract chromosome numbers and CDR start and end positions. For each chromosome, it adds the CDR regions to the `CDR_dict` dictionary. If a chromosome is not already in the dictionary, it initializes a new list for that chromosome. If the chromosome is already present, it appends the new CDR region to the existing list.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2de5652f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "''' here in the code, I am formulating the CDR regions and listing the CDRs in each and every chromosome'''\n",
    "input_file = '/private/groups/migalab/dan/data_analysis/young_old_analysis/HG002_DiMeLo_CENPA_youngpassage.hmmCDR_only_CDR_Dan_certified.bed'\n",
    "CDR_dict = {}\n",
    "with open(input_file, 'r') as infile:  \n",
    "    for i in infile:\n",
    "        chr_num = i.split('\\t')[0]\n",
    "        CDR_start = int (i.split('\\t')[1]) \n",
    "        CDR_end = int(i.split('\\t')[2].split('\\n')[0]) \n",
    "        if chr_num not in CDR_dict:\n",
    "            CDR_dict[chr_num] = [[CDR_start ,CDR_end]]\n",
    "        elif chr_num in CDR_dict:  \n",
    "            CDR_dict[chr_num].append ([CDR_start ,CDR_end])\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcc460e3",
   "metadata": {},
   "source": [
    "### Inputs:\n",
    "- `CDR_dict`: Dictionary where chromosome names are keys and the values are lists of CDR regions represented by their start and end positions.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': [[1000, 2000], [3000, 4000]],\n",
    "        'chr2': [[500, 1500]]\n",
    "    }\n",
    "    ```\n",
    "\n",
    "### Outputs:\n",
    "- `CDR_adjacent`: Dictionary where chromosome names are keys and the values are lists of CDR adjacent regions, with each adjacent region being 1000 base pairs long.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': [[-1, 999], [2001, 3001], [2999, 3999], [4001, 5001]],\n",
    "        'chr2': [[-501, 499], [1501, 2501]]\n",
    "    }\n",
    "    ```\n",
    "\n",
    "### Description:\n",
    "This code creates a dictionary, `CDR_adjacent`, to store the regions adjacent to the CDR (Conserved Domain Region) regions for each chromosome. For each chromosome in the `CDR_dict` dictionary, it iterates over the CDR regions and calculates the left and right adjacent regions, each 1000 base pairs long. These adjacent regions are then added to the `CDR_adjacent` dictionary under the corresponding chromosome key. The resulting dictionary provides information on the genomic regions immediately adjacent to each CDR region.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72fe3916",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Based on the CDR regions, I am obtaining CDR adjacent regions in the same format as the CDR data set above'''\n",
    "CDR_adjacent = {}\n",
    "for chromosome in CDR_dict: \n",
    "    CDR_adjacent[chromosome] =[]\n",
    "    for CDR in CDR_dict[chromosome]: \n",
    "        #print (CDR) \n",
    "        CDR_adjacent_left_space = [int(CDR[0]) - 1001, int(CDR[0]) - 1]\n",
    "        CDR_adjacent_right_space = [int(CDR[1]) + 1, int(CDR[1]) + 1001]\n",
    "        \n",
    "        CDR_adjacent[chromosome].append (CDR_adjacent_left_space)\n",
    "        CDR_adjacent[chromosome].append (CDR_adjacent_right_space)\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8720da8",
   "metadata": {},
   "source": [
    "### Inputs:\n",
    "- `CDR_dict`: Dictionary where chromosome names are keys and the values are lists of CDR regions represented by their start and end positions.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': [[1000, 2000], [3000, 4000]],\n",
    "        'chr2': [[500, 1500]]\n",
    "    }\n",
    "    ```\n",
    "- `active_dict`: Dictionary where chromosome names are keys and the values are lists of active array blocks represented by their start and end positions.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': [[0, 100000]],\n",
    "        'chr2': [[0, 100000]]\n",
    "    }\n",
    "    ```\n",
    "\n",
    "### Outputs:\n",
    "- `none_CDR_active`: Dictionary where chromosome names are keys and the values are lists of non-CDR active regions represented by their start and end positions.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': [[0, 999], [2001, 2999], [4001, 99999]],\n",
    "        'chr2': [[0, 499], [1501, 99999]]\n",
    "    }\n",
    "    ```\n",
    "\n",
    "### Description:\n",
    "This code generates a dictionary, `none_CDR_active`, to store non-CDR active regions for each chromosome. It uses the `CDR_dict` and `active_dict` dictionaries to determine the regions that are active but not part of the CDRs. For each chromosome, it calculates the regions from the start of the active block to the start of the first CDR, the regions between consecutive CDRs, and the regions from the end of the last CDR to the end of the active block. These non-CDR regions are added to the `none_CDR_active` dictionary under the corresponding chromosome key. The resulting dictionary provides information on the active genomic regions that are not part of any CDR.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e1d4593",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "none_CDR_active = {}\n",
    "for chromosome in CDR_dict: \n",
    "    none_CDR_active[chromosome] = [] \n",
    "    active_start = active_dict[chromosome][0][0]\n",
    "    active_end = active_dict[chromosome][0][1]\n",
    "    CDR_start_position = int (CDR_dict[chromosome][0][0]) - 1 \n",
    "    CDR_end_position = int (CDR_dict[chromosome][-1][1]) + 1\n",
    "    none_CDR_active[chromosome].append ([active_start, CDR_start_position])\n",
    "    for coordinate in range(len(CDR_dict[chromosome]) - 1):\n",
    "        # Generate in-between coordinates\n",
    "        \n",
    "        start_of_next = int (CDR_dict[chromosome][coordinate + 1][0]) - 1 \n",
    "        end_of_current = int (CDR_dict[chromosome][coordinate][1]) + 1 \n",
    "        # Ensure there's no overlap and the next start is greater than the current end\n",
    "        if int(start_of_next) > int(end_of_current):\n",
    "            none_CDR_active[chromosome].append([end_of_current, start_of_next])\n",
    "    \n",
    "    none_CDR_active[chromosome].append ([CDR_end_position, active_end])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c01034b",
   "metadata": {},
   "source": [
    "### Inputs:\n",
    "- `segment_num`: Number of random segments to pick.\n",
    "- `chr_name`: Chromosome name for which the random regions are to be generated.\n",
    "- `H1L_active_dict`: Dictionary where chromosome names are keys and the values are lists of active array blocks represented by their start and end positions.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': [[0, 100000]],\n",
    "        'chr2': [[0, 100000]]\n",
    "    }\n",
    "    ```\n",
    "- `assembly_sequence_length`: Dictionary containing chromosome names as keys and their sequence lengths as values.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': 248956422,\n",
    "        'chr2': 242193529\n",
    "    }\n",
    "    ```\n",
    "\n",
    "### Outputs:\n",
    "- `chromosome_arm_random_region_dict`: Dictionary where chromosome names are keys and the values are lists of randomly picked chromosome arm regions represented by their start and end positions.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': [[0, 99999], [200000, 299999], [400000, 499999]],\n",
    "        'chr2': [[0, 99999], [200000, 299999], [400000, 499999]]\n",
    "    }\n",
    "    ```\n",
    "\n",
    "### Description:\n",
    "This code defines a function `chromosome_arm_random_region` to generate random regions in the chromosome arms for a specified chromosome. It first calculates the active array regions for the chromosome and determines the excluded portions. Then, it randomly selects a defined number of segments from the non-active regions, ensuring no duplicate random numbers are chosen. The selected regions are expanded to chromosome positions and added to a list. The function returns a list of randomly picked regions for the chromosome.\n",
    "\n",
    "The code then creates a dictionary, `chromosome_arm_random_region_dict`, by applying the `chromosome_arm_random_region` function to each chromosome in `CDR_dict`. The resulting dictionary stores the randomly picked chromosome arm regions for each chromosome.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed306fe3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def chromosome_arm_random_region (segment_num, chr_name, H1L_active_dict, assembly_sequence_length): \n",
    "    # create variables to contain regions chosen \n",
    "    chromosome_arm_regions = []\n",
    "    excluded_portion = []\n",
    "    H1L_active_length = 0 \n",
    "    \n",
    "    for active_region in H1L_active_dict[chr_name]: \n",
    "        #calculate the length of the active array region for each chromosome \n",
    "        H1L_active_length = active_region[1] - active_region[0]\n",
    "        \n",
    "        #defining where the chromosome arm regions are\n",
    "        \n",
    "        #The left side of the active array \n",
    "        chromosome_arm_regions.append([0,active_region[0] - 1])\n",
    "        \n",
    "        #The right side of the active array \n",
    "        chromosome_arm_regions.append([active_region[1] + 1,assembly_sequence_length[chr_name]])\n",
    "        \n",
    "        #calculate the percentage portion of where the active array is in and add them to the exclusion bin \n",
    "        start_portion = active_region[0] / H1L_active_length\n",
    "        end_portion = active_region[1] / H1L_active_length\n",
    "        excluded_portion.append (int(start_portion))\n",
    "        excluded_portion.append (int(end_portion))\n",
    "        \n",
    "    #calculate the total portions of the active array \n",
    "    total_segment_amount = int(assembly_sequence_length[chr_name] / H1L_active_length)\n",
    "    \n",
    "    \n",
    "    #pick defined amount of random numbers between 0 and the pre defined amount of random numbers \n",
    "    random_numbers = []\n",
    "    for num in range(segment_num):\n",
    "        while True: \n",
    "            #if the same random number gets picked twice, repeat \n",
    "            current_random_number = random.randint(0, total_segment_amount)\n",
    "            if current_random_number not in (excluded_portion and random_numbers):\n",
    "                break\n",
    "            \n",
    "            \n",
    "        random_numbers.append(current_random_number)\n",
    "    \n",
    "    #Expand the chromosome portion number to chromosome position number by multiplying \n",
    "    random_picked_regions = [num * H1L_active_length for num in random_numbers]\n",
    "    uncoded_region_list = []\n",
    "    \n",
    "    #Make a dictionary that contains randomly picked region for each chromosome \n",
    "    for item in random_picked_regions: \n",
    "        arms_region_start = item\n",
    "        arms_region_end = item + H1L_active_length\n",
    "        uncoded_region_list.append([arms_region_start, arms_region_end])\n",
    "        \n",
    "    return uncoded_region_list\n",
    "\n",
    "\n",
    "chromosome_arm_random_region_dict = {}\n",
    "for chromosome in CDR_dict: \n",
    "    chromosome_arm_random_region_dict[chromosome] = chromosome_arm_random_region (3,\n",
    "                                                                             chromosome, \n",
    "                                                                             active_dict, \n",
    "                                                                             assembly_sequence_length)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43300dfa",
   "metadata": {},
   "source": [
    "### Inputs:\n",
    "- `cdr_adjacent`: Dictionary containing chromosome names as keys and lists of regions as values. The regions are represented by their start and end positions.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': [[-1, 999], [2001, 3001]],\n",
    "        'chr2': [[-501, 499], [1501, 2501]]\n",
    "    }\n",
    "    ```\n",
    "\n",
    "### Outputs:\n",
    "- `filtered_cdr_adjacent`: Dictionary with the same structure as the input but with invalid regions removed. Regions where the end position is not greater than the start position are excluded.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': [[2001, 3001]],\n",
    "        'chr2': [[1501, 2501]]\n",
    "    }\n",
    "    ```\n",
    "\n",
    "### Description:\n",
    "The function `check_lists` filters out invalid regions from a dictionary of chromosome regions. For each chromosome, it iterates through the list of regions and checks if the end position of each region is greater than its start position. If a region is invalid, it is excluded, and a message is printed indicating the removal of the invalid region. The function returns a new dictionary with only valid regions.\n",
    "\n",
    "The `check_lists` function can be used to validate and filter regions in various dictionaries, such as `CDR_dict`, `CDR_adjacent`, `none_CDR_active`, and `chromosome_arm_random_region_dict`, ensuring that only valid regions are retained for further analysis.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d95e2d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_lists(cdr_adjacent):\n",
    "    filtered_cdr_adjacent = {}\n",
    "    \n",
    "    for chromosome, regions in cdr_adjacent.items():\n",
    "        valid_regions = []\n",
    "        for region in regions:\n",
    "            if region[1] > region[0]:\n",
    "                valid_regions.append(region)\n",
    "            else:\n",
    "                print(f\"Removed invalid region in {chromosome}: {region[1]} is not larger than {region[0]}\")\n",
    "        if valid_regions:\n",
    "            filtered_cdr_adjacent[chromosome] = valid_regions\n",
    "\n",
    "    return filtered_cdr_adjacent\n",
    "CDR_dict= check_lists(CDR_dict)\n",
    "CDR_adjacent = check_lists(CDR_adjacent)\n",
    "none_CDR_active = check_lists(none_CDR_active)\n",
    "chromosome_arm_random_region_dict = check_lists(chromosome_arm_random_region_dict)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f29051bc",
   "metadata": {},
   "source": [
    "### Inputs:\n",
    "- `mod_no_dash`: Numpy array of the mod without any insertions and deletions.\n",
    "- `alignment_dash`: Alignment sequence with dashes (insertions).\n",
    "- `target_start_no_dash`: The starting position of the subset without dashes.\n",
    "- `target_end_no_dash`: The ending position of the subset without dashes.\n",
    "\n",
    "### Outputs:\n",
    "- `mod_subset`: Numpy array representing the subset of the mod without any insertions and deletions.\n",
    "\n",
    "### Description:\n",
    "The function `mod_subset_producing_step` isolates the desired regions (subset) from a numpy array of mod sequences (`mod_no_dash`) without insertions and deletions. It takes an alignment sequence with dashes (`alignment_dash`) and target start and end positions without dashes. The function creates a mask to identify non-dash positions, generates cumulative counts for non-dash positions, and maps these counts to the positions in the original alignment sequence.\n",
    "\n",
    "The function calculates the start and end positions of the target region within the dashed alignment sequence. It then extracts the subset of the alignment sequence, removes dashes, and calculates the corresponding start and end positions within the mod numpy array. The resulting subset of the mod numpy array (`mod_subset`) is returned, representing the desired region without any insertions or deletions.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e948347",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "The idea of this function is to isolate the the desired regions (here in the function, it is called the subset) in the mod \n",
    "numpy array without dashes(insertions)'''\n",
    "\n",
    "def mod_subset_producing_step (mod_no_dash,alignment_dash,target_start_no_dash,target_end_no_dash):\n",
    "    #mod_no_dash = is the numpy array of the mod without any insertions and deletions\n",
    "    # alignment_dash = is the alignment sequence with the dashes in it \n",
    "    # target_start = it's the subset starting position WITHOUT the dashes!!! \n",
    "\n",
    "\n",
    "    # Create a mask to identify non-dash positions\n",
    "    mask = [char != '-' for char in alignment_dash]\n",
    "\n",
    "    # Generate cumulative counts only for True values in the mask\n",
    "    cumulative_counts = list(itertools.accumulate(mask))\n",
    "\n",
    "\n",
    "    \n",
    "    # Create the final indexes list\n",
    "    indexes = [count - 1 if is_non_dash else '-' for count, is_non_dash in zip(cumulative_counts, mask)]\n",
    "\n",
    "\n",
    "\n",
    "    target_start_dash = indexes.index (target_start_no_dash)\n",
    "\n",
    "        \n",
    "    try:\n",
    "        target_end_dash = indexes.index (target_end_no_dash)\n",
    "    except ValueError: \n",
    "        target_end_dash = indexes[-1]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    #obtain dashed alignment \n",
    "    alignment_dash_sequence_pre_subset = alignment_dash[0:target_start_dash]\n",
    "    alignment_dash_sequence_subset = alignment_dash[target_start_dash:target_end_dash]\n",
    "\n",
    "    #create no dash alignment \n",
    "    alignment_no_dash_sequence_pre_subset = alignment_dash_sequence_pre_subset.replace(\"-\",\"\")\n",
    "    alignment_no_dash_sequence_subset = alignment_dash_sequence_subset.replace(\"-\",\"\")\n",
    "\n",
    "    subset_no_dash_start = len(alignment_no_dash_sequence_pre_subset)\n",
    "    subset_no_dash_end = subset_no_dash_start + len(alignment_no_dash_sequence_subset)\n",
    "\n",
    "    #make mod_no_dash alignment\n",
    "    mod_subset = mod_no_dash[subset_no_dash_start:subset_no_dash_end]\n",
    "\n",
    "    return mod_subset\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63cb6acb",
   "metadata": {},
   "source": [
    "### Inputs:\n",
    "- `chromosome_coordinates`: Dictionary containing chromosome names as keys and lists of regions represented by their start and end positions.\n",
    "  - Example:\n",
    "    ```python\n",
    "    {\n",
    "        'chr1': [[1000, 2000], [3000, 4000]],\n",
    "        'chr2': [[500, 1500]]\n",
    "    }\n",
    "    ```\n",
    "- `name`: Name identifier for the output file.\n",
    "- `bamfile`: BAM file object containing read data.\n",
    "- `mod_tag`: Modification tag to calculate the density for (e.g., 'A' for adenine).\n",
    "\n",
    "### Outputs:\n",
    "- `data_table`: List of lists containing chromosome name, region, average region density, and normalized region base coverage.\n",
    "  - Example:\n",
    "    ```python\n",
    "    [\n",
    "        ['chr1', [1000, 2000], 0.123456, 0.987654],\n",
    "        ['chr2', [500, 1500], 0.654321, 0.876543]\n",
    "    ]\n",
    "    ```\n",
    "- CSV file: Saved to the specified directory with region density scores.\n",
    "\n",
    "### Description:\n",
    "The `region_read_mA_density_calculator` function calculates the density of modifications (e.g., methylation) in specified regions of chromosomes. It takes a dictionary of chromosome coordinates, a name identifier, a BAM file, and a modification tag. For each region in each chromosome, the function fetches reads from the BAM file, aligns them, and calculates the modification density. It handles different scenarios where regions and reads overlap in various ways, ensuring that only valid reads and regions are processed. The function then compiles the data into a table and saves it as a CSV file. It also prints the name identifier and the formatted table.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f621083c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def region_read_mA_density_calculator (chromosome_coordinates,name,bamfile,mod_tag): \n",
    "    data_table = [] \n",
    "\n",
    "    #get each chromosome\n",
    "    for chr_name in chromosome_coordinates:\n",
    "        for region in chromosome_coordinates[chr_name]:\n",
    "            region_density = []\n",
    "            region_base = 0 \n",
    "            \n",
    "            region_start_index = int(region[0])\n",
    "            region_end_index = int(region[1])\n",
    "\n",
    "            \n",
    "            for read in bamfile.fetch(chr_name,region_start_index,region_end_index):\n",
    "\n",
    "                #make an if statement to check a specific read front, middle, end regions \n",
    "                #setting read start, end, density, length variables \n",
    "                    \n",
    "                #Get the starting and ending positions of the reads \n",
    "                read_start_position = read.reference_start\n",
    "                read_end_position = read.reference_end\n",
    "                read_density = 0 \n",
    "        \n",
    "                #Get sequence information which shows deletions and insertions \n",
    "                sequence = read.get_aligned_pairs(matches_only=False, with_seq = True)\n",
    "\n",
    "\n",
    "                #make a numpy of the sequence length which eliminates the deletion\n",
    "                \n",
    "                read_sequence_insertion_included = ''\n",
    "                genomic_alignment_sequence_deletion_mistach_included = ''\n",
    "                \n",
    "                for item in sequence:\n",
    "                    if item[0] is None:\n",
    "                        read_sequence_insertion_included+='-'\n",
    "                    elif item[1] is None:\n",
    "                        genomic_alignment_sequence_deletion_mistach_included += '-'\n",
    "                    else: \n",
    "                        read_sequence_insertion_included+=item[2]\n",
    "                        genomic_alignment_sequence_deletion_mistach_included +=item[2]\n",
    "\n",
    "                \n",
    "                read_sequence_insertion_included = read_sequence_insertion_included.upper()\n",
    "                genomic_alignment_sequence_deletion_mistach_included = genomic_alignment_sequence_deletion_mistach_included.upper()\n",
    "\n",
    "\n",
    "                genomic_alignment_sequence_deletion_mistach_included_mask = np.array(\n",
    "                    [char != '-' for char in genomic_alignment_sequence_deletion_mistach_included])\n",
    "\n",
    "                #take sequence length excluding insertions \n",
    "                insertions = read_sequence_insertion_included.count (\"-\")\n",
    "                no_insertion_no_deletion_sequence_length = len(read_sequence_insertion_included) \n",
    "                \n",
    "                \n",
    "                \n",
    "                #removing reads shorter than 50000 \n",
    "                #if no_insertion_no_deletion_sequence_length < 50000:\n",
    "                #    continue \n",
    "\n",
    "                #make a mod np array with the length of the read length\n",
    "                mod=read.modified_bases_forward\n",
    "                \n",
    "                #make a mod score with its original length \n",
    "                mod_score = np.zeros(len(genomic_alignment_sequence_deletion_mistach_included),)\n",
    "\n",
    "                \n",
    "                #make transfer mA positions to mod np array corresponded to their sequence positions \n",
    "                try:\n",
    "                    if mod_tag == 'A':\n",
    "                        for indices, values in mod[('A', 0, 'a')]:\n",
    "                            mod_score[indices] = values\n",
    "                        mod_score = mod_score[genomic_alignment_sequence_deletion_mistach_included_mask]\n",
    "                        \n",
    "\n",
    "\n",
    "                    elif mod_tag == 'CG':\n",
    "                        for indices, values in mod[('C', 0, 'm')]:\n",
    "                            mod_score[indices] = values\n",
    "                        mod_score = mod_score[genomic_alignment_sequence_deletion_mistach_included_mask]\n",
    "                    \n",
    "                    if read.is_reverse:\n",
    "                            mod_score = mod_score[::-1]\n",
    "\n",
    "\n",
    "                # No mod would return KeyError \n",
    "                except KeyError:\n",
    "                    continue\n",
    "\n",
    "                \n",
    "                    \n",
    "\n",
    "\n",
    "                # if the regions are longer than the reads \n",
    "                if (region_end_index - region_start_index) > (read_end_position - read_start_position):\n",
    "                    # scenario 4: if the reads are inside the region\n",
    "                    if (region_end_index >= read_end_position) and (region_start_index <= read_start_position): \n",
    "                        mod_start = 0\n",
    "                        mod_end = len(read_sequence_insertion_included)\n",
    "                    \n",
    "                    # scenario 5: if the reads cover the later part of the region\n",
    "                    elif (region_end_index < read_end_position) and (region_start_index > read_start_position): \n",
    "                        mod_start = 0\n",
    "                        mod_end = no_insertion_no_deletion_sequence_length - read_end_position - region_end_index\n",
    "\n",
    "                    # scenario 6: if the reads cover the starting part of the region \n",
    "                    elif (region_end_index > read_end_position) and (region_start_index > read_start_position): \n",
    "                        mod_start = region_start_index - read_start_position \n",
    "                        mod_end = no_insertion_no_deletion_sequence_length\n",
    "\n",
    "                        \n",
    "                \n",
    "                # if the reads are longer than the region selected \n",
    "                elif (region_end_index - region_start_index) <= (read_end_position - read_start_position):\n",
    "                    # scenario 1: when the defined region is inside the read\n",
    "                    if (read_start_position <= region_start_index) and (read_end_position >= region_end_index):\n",
    "                        mod_start = region_start_index - read_start_position \n",
    "                        mod_end = region_end_index - read_start_position\n",
    "\n",
    "                    # scenario 3: when the defined region covers a bit of the end of the read\n",
    "                    elif (read_end_position < region_end_index) and (read_end_position > region_start_index):\n",
    "                        mod_start = region_start_index - read_start_position\n",
    "\n",
    "                        mod_end = no_insertion_no_deletion_sequence_length\n",
    "\n",
    "                    # scenario 2: when the defined region covers a bit of the beginning of the read\n",
    "                    elif (read_start_position > region_start_index) and (read_start_position < region_end_index):\n",
    "                        mod_start = 0\n",
    "                        mod_end = region_end_index - read_start_position \n",
    "\n",
    "\n",
    "                #use the defined starting and ending positons in the region to subset mod numpy\n",
    "                if (region_start_index - read_start_position) > (no_insertion_no_deletion_sequence_length - insertions):\n",
    "                    continue\n",
    "                try:\n",
    "                    trimmed_mod_score = mod_subset_producing_step (mod_score,read_sequence_insertion_included,mod_start,mod_end)\n",
    "                except ValueError:\n",
    "                    continue\n",
    "                \n",
    "            \n",
    "                region_base += (mod_end - mod_start)\n",
    "                #removing all the zeros \n",
    "                mod_no_zeros = trimmed_mod_score[trimmed_mod_score != 0]\n",
    "                m_mod_tag = len (mod_no_zeros)\n",
    "                \n",
    "\n",
    "                #Getting the total amount of As in the subsetted region of the sequence \n",
    "                total_mod_tag = read_sequence_insertion_included[mod_start:mod_end].count(mod_tag)\n",
    "                \n",
    "                \n",
    "                #calculate read density\n",
    "                try:\n",
    "                    read_density = m_mod_tag / total_mod_tag\n",
    "                    \n",
    "                except ZeroDivisionError:\n",
    "                    pass\n",
    "                region_density.append (read_density)\n",
    "\n",
    "                \n",
    "                    \n",
    "                \n",
    "            #calculate averaged region density average \n",
    "            try:\n",
    "                region_density_average = sum(region_density)/len(region_density)\n",
    "            \n",
    "            except ZeroDivisionError:\n",
    "                region_density_average = 0\n",
    "            data_table.append ([chr_name,region,region_density_average,region_base/(region_end_index - region_start_index)])\n",
    "            #print (chr_name, region_end_index - region_start_index, region_base )\n",
    "\n",
    "        \n",
    "    table = tabulate(data_table, headers=\"firstrow\", tablefmt=\"fancy_grid\", floatfmt=\".18f\")\n",
    "    \n",
    "    filename = f\"/private/groups/migalab/dan/data_analysis/young_old_analysis/{name}_region_density_scores_{mod_tag}.csv\"\n",
    "    with open(filename, \"w\", newline=\"\") as csvfile:\n",
    "        writer = csv.writer(csvfile,delimiter=\"\\t\")\n",
    "        writer.writerows(data_table)\n",
    "\n",
    "    print (name)\n",
    "    print (table)\n",
    "    \n",
    "\n",
    "#region_read_mA_density_calculator(test_dict,'test_dict',bamfile)\n",
    "\n",
    "#for scenario in scenarios: \n",
    "    #region_read_mA_density_calculator (scenario,str(scenario))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4651219a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "import unittest\n",
    "from unittest.mock import MagicMock, patch\n",
    "class TestRegionReadMADensityCalculator(unittest.TestCase):\n",
    "\n",
    "    @patch('pysam.AlignmentFile')\n",
    "    def test_region_read_mA_density_calculator(self, mock_bamfile):\n",
    "        # Mock chromosome names\n",
    "        mock_bamfile.references = ['chr1']\n",
    "        \n",
    "        # Define mock reads and return values\n",
    "        mock_reads = []\n",
    "        for i in range(100):\n",
    "            mock_read = MagicMock()\n",
    "            mock_read.reference_start = 100\n",
    "            mock_read.reference_end = 500\n",
    "            # Create modified bases with mA every other A position\n",
    "\n",
    "            mock_read.modified_bases_forward= {('A', 0, 'a'): [(0, 254), (4, 254), (8, 254),(16, 254), (24, 254),(32, 254), (36, 254), (40, 254),(48, 254), (52, 254)]\n",
    "                                               +[(104, 254), (108, 254), (112, 254),(116, 254), (152, 254)]\n",
    "                                               +[(j, 250) for j in range(200, 300, 4)]\n",
    "                                               +[(396, 254)]\n",
    "                                               }\n",
    "            \n",
    "            # Mock the sequence as a long string of 'A's\n",
    "            mock_read.get_aligned_pairs.return_value = [(j, j, 'A') for j in range(0, 400)]\n",
    "            \n",
    "            mock_read.is_reverse = False\n",
    "            \n",
    "            mock_reads.append(mock_read)\n",
    "\n",
    "        # Mock the fetch method\n",
    "        mock_bamfile.fetch.side_effect = lambda chr_name, start, end: iter(mock_reads) if chr_name == 'chr1' else iter([])\n",
    "\n",
    "        # Define input\n",
    "        chromosome_coordinates = {'chr1': [[100, 200]]}\n",
    "        name = 'test_output'\n",
    "        mod_tag = 'A'\n",
    "        \n",
    "        # Call the function\n",
    "        region_read_mA_density_calculator(chromosome_coordinates, name, mock_bamfile,mod_tag)\n",
    "        \n",
    "        # Check if the file was created and contains the expected data\n",
    "        with open('test_output_region_density_scores.csv', 'r') as file:\n",
    "            content = file.read()\n",
    "        \n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    unittest.TextTestRunner().run(unittest.TestLoader().loadTestsFromTestCase(TestRegionReadMADensityCalculator))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2a9d2a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import unittest\n",
    "from unittest.mock import MagicMock, patch\n",
    "class TestRegionReadMADensityCalculator(unittest.TestCase):\n",
    "\n",
    "    @patch('pysam.AlignmentFile')\n",
    "    def test_region_read_mA_density_calculator(self, mock_bamfile):\n",
    "        # Mock chromosome names\n",
    "        mock_bamfile.references = ['chr1']\n",
    "        \n",
    "        # Define mock reads and return values\n",
    "        mock_reads = []\n",
    "        for i in range(100):\n",
    "            mock_read = MagicMock()\n",
    "            mock_read.reference_start = 100\n",
    "            mock_read.reference_end = 600\n",
    "            # Create modified bases with mA every other A position\n",
    "\n",
    "            mock_read.modified_bases_forward= {('A', 0, 'a'): \n",
    "                                               \n",
    "                                               [(j, 250) for j in range(0, 125)]\n",
    "                                               +[(j, 250) for j in range(126, 227)]\n",
    "                                               +[(j, 250) for j in range(251, 326)]\n",
    "                                               +[(j, 250) for j in range(376, 411)]\n",
    "                                               }\n",
    "            \n",
    "            # Mock the sequence as a long string of 'A's\n",
    "\n",
    "            # Initialize the list to store the tuples\n",
    "            mock_sequence = []\n",
    "\n",
    "            # Initialize the second position counter\n",
    "            second_pos = 0\n",
    "\n",
    "            # Generate the tuples\n",
    "            for i in range(500):\n",
    "                if i % 5 == 0:\n",
    "                    mock_sequence.append((i, None, None))\n",
    "                else:\n",
    "                    mock_sequence.append((i, second_pos, 'A'))\n",
    "                    second_pos += 1\n",
    "\n",
    "                # Reset the second position counter if it exceeds 400\n",
    "                if second_pos > 400:\n",
    "                    second_pos = 0\n",
    "            mock_read.get_aligned_pairs.return_value = mock_sequence\n",
    "            \n",
    "            mock_read.is_reverse = False\n",
    "            \n",
    "            mock_reads.append(mock_read)\n",
    "\n",
    "        # Mock the fetch method\n",
    "        mock_bamfile.fetch.side_effect = lambda chr_name, start, end: iter(mock_reads) if chr_name == 'chr1' else iter([])\n",
    "\n",
    "        # Define input\n",
    "        chromosome_coordinates = {'chr1': [[400, 500]]}\n",
    "        name = 'test_output'\n",
    "        \n",
    "        # Call the function\n",
    "        region_read_mA_density_calculator(chromosome_coordinates, name, mock_bamfile,'A')\n",
    "        \n",
    "        # Check if the file was created and contains the expected data\n",
    "        with open('test_output_region_density_scores.csv', 'r') as file:\n",
    "            content = file.read()\n",
    "        \n",
    "\n",
    "if __name__ == '__main__':\n",
    "    unittest.TextTestRunner().run(unittest.TestLoader().loadTestsFromTestCase(TestRegionReadMADensityCalculator))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6806c86",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import unittest\n",
    "from unittest.mock import MagicMock, patch\n",
    "class TestRegionReadMADensityCalculator(unittest.TestCase):\n",
    "\n",
    "    @patch('pysam.AlignmentFile')\n",
    "    def test_region_read_mA_density_calculator(self, mock_bamfile):\n",
    "        # Mock chromosome names\n",
    "        mock_bamfile.references = ['chr1']\n",
    "        \n",
    "        # Define mock reads and return values\n",
    "        mock_reads = []\n",
    "        for i in range(100):\n",
    "            mock_read = MagicMock()\n",
    "            mock_read.reference_start = 100\n",
    "            mock_read.reference_end = 500\n",
    "            # Create modified bases with mA every other A position\n",
    "\n",
    "            mock_read.modified_bases_forward= {('A', 0, 'a'): \n",
    "                                               \n",
    "                                               [(i, 250) if (i + 1) % 4 != 0 else (i, 0) for i in range(0,133)]\n",
    "                                               +[(i, 250) if (i + 1) % 4 != 0 else (i, 0) for i in range(134,202)]\n",
    "\n",
    "                                               \n",
    "                                             \n",
    "                                               }\n",
    "            \n",
    "            # Mock the sequence as a long string of 'A's\n",
    "\n",
    "            # Initialize the list to store the tuples\n",
    "            # Initialize the list to store the tuples\n",
    "            mock_sequence = []\n",
    "\n",
    "            # Initialize the counters for the first and second positions\n",
    "            first_pos = 0\n",
    "            second_pos = 100\n",
    "\n",
    "            # Generate the tuples\n",
    "            for i in range(400):\n",
    "                if (i + 1) % 4 == 0:\n",
    "                    mock_sequence.append((None, second_pos, 'A'))\n",
    "                else:\n",
    "                    mock_sequence.append((first_pos, second_pos, 'A'))\n",
    "                    first_pos += 1\n",
    "                \n",
    "                second_pos += 1\n",
    "\n",
    "\n",
    "\n",
    "            mock_read.get_aligned_pairs.return_value = mock_sequence\n",
    "            \n",
    "            mock_read.is_reverse = False\n",
    "            \n",
    "            mock_reads.append(mock_read)\n",
    "\n",
    "        # Mock the fetch method\n",
    "        mock_bamfile.fetch.side_effect = lambda chr_name, start, end: iter(mock_reads) if chr_name == 'chr1' else iter([])\n",
    "\n",
    "        # Define input\n",
    "        chromosome_coordinates = {'chr1': [[233, 366]]}\n",
    "        name = 'test_output'\n",
    "        \n",
    "        # Call the function\n",
    "        region_read_mA_density_calculator(chromosome_coordinates, name, mock_bamfile)\n",
    "        \n",
    "        # Check if the file was created and contains the expected data\n",
    "        with open('test_output_region_density_scores.csv', 'r') as file:\n",
    "            content = file.read()\n",
    "        \n",
    "\n",
    "if __name__ == '__main__':\n",
    "    unittest.TextTestRunner().run(unittest.TestLoader().loadTestsFromTestCase(TestRegionReadMADensityCalculator))\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2a61279",
   "metadata": {},
   "outputs": [],
   "source": [
    "CDR_dict= check_lists(CDR_dict)\n",
    "CDR_adjacent = check_lists(CDR_adjacent)\n",
    "none_CDR_active = check_lists(none_CDR_active)\n",
    "chromosome_arm_random_region_dict = check_lists(chromosome_arm_random_region_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aeb45b4f",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "80135d81",
   "metadata": {},
   "source": [
    "### Inputs:\n",
    "- `bar_plot_dataset`: List of dictionaries containing chromosome regions to analyze.\n",
    "  - Example:\n",
    "    ```python\n",
    "    [CDR_dict, none_CDR_active, CDR_adjacent]\n",
    "    ```\n",
    "- `bar_plot_dataset_names`: List of names corresponding to each dictionary in `bar_plot_dataset`.\n",
    "  - Example:\n",
    "    ```python\n",
    "    ['CENPA_young_CDR_dict', 'CENPA_young_none_CDR_active_dict', 'CDR_young_adjacent']\n",
    "    ```\n",
    "- `bamfile`: BAM file object containing read data.\n",
    "\n",
    "### Outputs:\n",
    "- Calls the `region_read_mA_density_calculator` function for each dictionary in `bar_plot_dataset` and saves the results to CSV files.\n",
    "\n",
    "### Description:\n",
    "This code snippet uses parallel processing to calculate the modification density for different sets of chromosome regions using the `region_read_mA_density_calculator` function. It iterates over the list of dictionaries (`bar_plot_dataset`), each containing different sets of chromosome regions, and their corresponding names (`bar_plot_dataset_names`). For each dictionary, it calls `region_read_mA_density_calculator`, passing the dictionary, its name, the BAM file, and the modification tag ('A'). The results are saved to CSV files named according to the dataset names.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "799e17d7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from joblib import Parallel, delayed\n",
    "\n",
    "bar_plot_dataset = [CDR_dict, none_CDR_active, CDR_adjacent]\n",
    "bar_plot_dataset_names = [ 'CENPA_young_CDR_dict', 'CENPA_young_none_CDR_active_dict','CDR_young_adjacent']\n",
    "def get_variable_name(var, locals_dict):\n",
    "    for name, value in locals_dict.items():\n",
    "        if value is var:\n",
    "            return name\n",
    "    return None\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "for dictionary in range (0, len (bar_plot_dataset)): \n",
    "\n",
    "    region_read_mA_density_calculator (bar_plot_dataset[dictionary],bar_plot_dataset_names[dictionary],bamfile,'A')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c72e0b4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "region_read_mA_density_calculator (CDR_dict,'halo_old_passaged_',bamfile,'A')\n",
    "#region_read_mA_density_calculator (CDR_dict,'halo_young_passaged_',bamfile,'CG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be09b8fe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "641cdd8b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4de075da",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a0c9764",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db06d0a0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6911d5aa",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
