{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 527,
   "id": "6b835e59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40.48095941543579 seconds\n"
     ]
    }
   ],
   "source": [
    "import pysam\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import time \n",
    "import matplotlib.patches as patches\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.pyplot as plt\n",
    "import itertools\n",
    "import pandas as pd \n",
    "from tabulate import tabulate\n",
    "import csv\n",
    "import random\n",
    "\n",
    "np.set_printoptions(threshold=np.inf)\n",
    "min_quality_score = 8\n",
    "\n",
    "#Load the bam file \n",
    "bamfile = pysam.AlignmentFile(\n",
    "    \"/Data2/01_09_24_R1041_DiMeLo_MsIgg/01_09_24_R1041_DiMeLo_MsIgg_5mC_6mA_winnowmap_sorted_sorted_clean_filtered.bam\",\n",
    "    \"rb\") \n",
    "\n",
    "\n",
    "assembly_ = open(\"/Data1/hg002v1.0.1.fasta\", \"r\")\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "#Load the reference genome and make it into a dictionary \n",
    "fasta_sequences = SeqIO.parse(assembly_, \"fasta\")\n",
    "assembly={}\n",
    "for fasta in fasta_sequences:\n",
    "    name, sequence = fasta.id, str(fasta.seq)\n",
    "    assembly[name] = sequence\n",
    "\n",
    "#Make a dictionary for all the chromosomes and their corresponding sequence length \n",
    "assembly_sequence_length = {}    \n",
    "for chromosome in assembly:\n",
    "    assembly_sequence_length[chromosome] = len(assembly[chromosome])\n",
    "    \n",
    "end_time = time.time()\n",
    "elapsed_time = end_time - start_time\n",
    "print (elapsed_time, \"seconds\")\n",
    "assembly_.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "cdba1098",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''The purpose of the code here is to make an active array where the chromosome name is\n",
    "the dictionary key. The active array blocks are in lists inside the key.\n",
    "'''\n",
    "import os\n",
    "input_file = '/Data1/HG002v0.1censat/hg002v1.0.fasta.manualAlpha.cenSat_H1L_merged.bed'\n",
    "active_dict = {}\n",
    "with open(input_file, 'r') as infile:  \n",
    "    num = 0 \n",
    "    previous_chr_num = \"\"\n",
    "    previous_parental_status = \"\"\n",
    "    for i in infile:\n",
    "        chr_num = i.split ('_')[0]\n",
    "        parental_status = i.split ('_')[1][0:8]\n",
    "        if chr_num == previous_chr_num and parental_status !=previous_parental_status:\n",
    "            num = 0 \n",
    "        elif chr_num == previous_chr_num and parental_status ==previous_parental_status:\n",
    "            pass \n",
    "        else: \n",
    "            num = 0 \n",
    "        active = i.split ()\n",
    "        if active[0] not in active_dict:\n",
    "            \n",
    "            active_dict[active[0]] = [[int(active[1]) ,int(active[2])]]\n",
    "        else:\n",
    "            active_dict[active[0]].append([int(active[1]) ,int(active[2])])\n",
    "        previous_chr_num = chr_num \n",
    "        previous_parental_status = parental_status\n",
    "#print (active_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "657d07bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "id": "2de5652f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "''' here in the code, I am formulating the CDR regions and listing the CDRs in each and every chromosome'''\n",
    "input_file = '/Data1/CDRStrictBedFiles/hg002v1.merged.strict.CDR.bed'\n",
    "CDR_dict = {}\n",
    "with open(input_file, 'r') as infile:  \n",
    "    for i in infile:\n",
    "        chr_num = i.split('\\t')[0]\n",
    "        CDR_start = i.split('\\t')[1]\n",
    "        CDR_end = i.split('\\t')[2].split('\\n')[0]\n",
    "        if chr_num not in CDR_dict:\n",
    "            CDR_dict[chr_num] = [[CDR_start ,CDR_end]]\n",
    "        elif chr_num in CDR_dict:  \n",
    "            CDR_dict[chr_num].append ([CDR_start ,CDR_end])\n",
    "\n",
    "#print (CDR_dict)\n",
    "'''This piece of code is to determine how many CDRs there are in a given chromosome'''\n",
    "for i in CDR_dict:\n",
    "#    print (i, len(CDR_dict[i]))\n",
    "    for CDR in CDR_dict[i]:\n",
    "        CDR_length = int(CDR[1]) - int(CDR[0])\n",
    "        total_A_count = assembly[i][int(CDR[0]):int(CDR[1])].count (\"A\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "id": "72fe3916",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Based on the CDR regions, I am obtaining CDR adjacent regions in the same format as the CDR data set above'''\n",
    "CDR_adjacent = {}\n",
    "for chromosome in CDR_dict: \n",
    "    CDR_adjacent[chromosome] =[]\n",
    "    for CDR in CDR_dict[chromosome]: \n",
    "        #print (CDR) \n",
    "        CDR_adjacent_left_space = [int(CDR[0]) - 1001, int(CDR[0]) - 1]\n",
    "        CDR_adjacent_right_space = [int(CDR[1]) + 1, int(CDR[1]) + 1001]\n",
    "        \n",
    "        CDR_adjacent[chromosome].append (CDR_adjacent_left_space)\n",
    "        CDR_adjacent[chromosome].append (CDR_adjacent_right_space)\n",
    "        \n",
    "    \n",
    "        \n",
    "#print (CDR_adjacent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "id": "6e1d4593",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'chrX_MATERNAL': [[59281767, 59282767], [59299768, 59300768], [59414767, 59415767], [59448768, 59449768], [59460767, 59461767], [59496768, 59497768]]}\n"
     ]
    }
   ],
   "source": [
    "chrom_X_CDR_dict ={}\n",
    "chrom_X_CDR_adjacent_dict = {}\n",
    "chrom_X_CDR_dict['chrX_MATERNAL'] = CDR_dict['chrX_MATERNAL']\n",
    "chrom_X_CDR_adjacent_dict ['chrX_MATERNAL'] = CDR_adjacent['chrX_MATERNAL']\n",
    "print (chrom_X_CDR_adjacent_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "id": "ed306fe3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def chromosome_arm_random_region (segment_num, chr_name, H1L_active_dict, assembly_sequence_length): \n",
    "    # create variables to contain regions chosen \n",
    "    chromosome_arm_regions = []\n",
    "    excluded_portion = []\n",
    "    H1L_active_length = 0 \n",
    "    \n",
    "    for active_region in H1L_active_dict[chr_name]: \n",
    "        #calculate the length of the active array region for each chromosome \n",
    "        H1L_active_length = active_region[1] - active_region[0]\n",
    "        \n",
    "        #defining where the chromosome arm regions are\n",
    "        \n",
    "        #The left side of the active array \n",
    "        chromosome_arm_regions.append([0,active_region[0] - 1])\n",
    "        \n",
    "        #The right side of the active array \n",
    "        chromosome_arm_regions.append([active_region[1] + 1,assembly_sequence_length[chr_name]])\n",
    "        \n",
    "        #calculate the percentage portion of where the active array is in and add them to the exclusion bin \n",
    "        start_portion = active_region[0] / H1L_active_length\n",
    "        end_portion = active_region[1] / H1L_active_length\n",
    "        excluded_portion.append (int(start_portion))\n",
    "        excluded_portion.append (int(end_portion))\n",
    "        \n",
    "    #calculate the total portions of the active array \n",
    "    total_segment_amount = int(assembly_sequence_length[chr_name] / H1L_active_length)\n",
    "    \n",
    "    \n",
    "    #pick defined amount of random numbers between 0 and the pre defined amount of random numbers \n",
    "    random_numbers = []\n",
    "    for num in range(segment_num):\n",
    "        while True: \n",
    "            #if the same random number gets picked twice, repeat \n",
    "            current_random_number = random.randint(0, total_segment_amount)\n",
    "            if current_random_number not in (excluded_portion and random_numbers):\n",
    "                break\n",
    "            \n",
    "            \n",
    "        random_numbers.append(current_random_number)\n",
    "    \n",
    "    #Expand the chromosome portion number to chromosome position number by multiplying \n",
    "    random_picked_regions = [num * H1L_active_length for num in random_numbers]\n",
    "    uncoded_region_list = []\n",
    "    \n",
    "    #Make a dictionary that contains randomly picked region for each chromosome \n",
    "    for item in random_picked_regions: \n",
    "        arms_region_start = item\n",
    "        arms_region_end = item + H1L_active_length\n",
    "        uncoded_region_list.append([arms_region_start, arms_region_end])\n",
    "        \n",
    "    return uncoded_region_list\n",
    "\n",
    "\n",
    "chromosome_arm_random_region_dict = {}\n",
    "for chromosome in chrom_X_CDR_dict: \n",
    "    chromosome_arm_random_region_dict[chromosome] = chromosome_arm_random_region (5,\n",
    "                                                                             chromosome, \n",
    "                                                                             active_dict, \n",
    "                                                                             assembly_sequence_length)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 524,
   "id": "2e948347",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'chrX_MATERNAL': [[37350684, 40463241], [6225114, 9337671], [143177622, 146290179], [68476254, 71588811]]}\n",
      "{'chrX_MATERNAL': [['59282768', '59299767'], ['59415768', '59448767'], ['59461768', '59496767']]}\n",
      "{'chrX_MATERNAL': [[59281767, 59282767], [59299768, 59300768], [59414767, 59415767], [59448768, 59449768], [59460767, 59461767], [59496768, 59497768]]}\n"
     ]
    }
   ],
   "source": [
    "#test_dict = {'chrX_MATERNAL':[['52000000','62000000']]}\n",
    "#scenario_1 = {'chrX_MATERNAL':[['51900000','51900810']]}\n",
    "#scenario_2 = {'chrX_MATERNAL':[['51900810','51901200']]}\n",
    "#scenario_3 = {'chrX_MATERNAL':[['52167000','52167900']]}\n",
    "#scenarios = [scenario_1,scenario_2,scenario_3]\n",
    "chromosome_arm_random_region_dict_chrX = {'chrX_MATERNAL': [[37350684, 40463241], [6225114, 9337671], [143177622, 146290179], [68476254, 71588811]]}\n",
    "print (chromosome_arm_random_region_dict_chrX)\n",
    "print (chrom_X_CDR_dict)\n",
    "print (chrom_X_CDR_adjacent_dict)\n",
    "chromX = [chrom_X_CDR_dict,chromosome_arm_random_region_dict_chrX,chrom_X_CDR_adjacent_dict]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 525,
   "id": "f621083c",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "''' \n",
    "The input of the function is a dictionary in the format of 'chromosome':[[start,end],[start,end]] \n",
    "'''\n",
    "def region_read_mA_density_calculator (chromosome_coordinates,name): \n",
    "    data_table = [] \n",
    "\n",
    "    #get each chromosome\n",
    "    for chr_name in chromosome_coordinates:\n",
    "\n",
    "        for region in chromosome_coordinates[chr_name]:\n",
    "            region_density = []\n",
    "\n",
    "            region_start_index = int(region[0])\n",
    "            region_end_index = int(region[1])\n",
    "            \n",
    "            for read in bamfile.fetch(chr_name,region_start_index,region_end_index):\n",
    "                #make an if statement to check a specific read front, middle, end regions \n",
    "                #setting read start, end, density, length variables \n",
    "                    \n",
    "                #Get the starting and ending positions of the reads \n",
    "                read_start_position = read.reference_start\n",
    "                read_end_position = read.reference_end\n",
    "                read_density = 0 \n",
    "        \n",
    "                #Get sequence information which shows deletions and insertions \n",
    "                sequence = read.get_aligned_pairs(matches_only=False, with_seq = True)\n",
    "\n",
    "                #make a numpy of the sequence length which eliminates the deletion\n",
    "                sequence_insertions = []\n",
    "                sequence_deletions = [] \n",
    "                \n",
    "                true_sequence = ''\n",
    "                \n",
    "                for item in sequence:\n",
    "                    if item[0] is None:\n",
    "                        sequence_deletions.append (item[1] - read_start_position)\n",
    "                    elif item[1] is None:\n",
    "                        sequence_insertions.append (item[0])\n",
    "                    else: \n",
    "                        true_sequence+=item[2]\n",
    "                \n",
    "                #take sequence length excluding insertions \n",
    "                insertion_amount = sum(1 for item in sequence if item[1] is None)\n",
    "                sequence_length = len(sequence) - insertion_amount\n",
    "                \n",
    "                #removing reads shorter than 50000 \n",
    "                if sequence_length < 50000:\n",
    "                    continue \n",
    "\n",
    "                #make a mod np array with the length of the read length\n",
    "                mod=read.modified_bases_forward\n",
    "                \n",
    "                #make a mod score with its original length \n",
    "                mod_score = np.zeros(len(read.query_sequence),)\n",
    "                \n",
    "                #make transfer mA positions to mod np array corresponded to their sequence positions \n",
    "                try:\n",
    "                    for indices, values in mod[('A', 0, 'a')]:\n",
    "                        mod_score[indices] = values\n",
    "\n",
    "                # No mod would return KeyError \n",
    "                except KeyError:\n",
    "                    continue\n",
    "                    \n",
    "                \n",
    "                #remove insertions \n",
    "\n",
    "                mod_score_insertions_removed = np.delete(mod_score, sequence_insertions)\n",
    "\n",
    "                #insert deletions \n",
    "                if len(sequence_deletions) == 0: \n",
    "                    mod_score_deletions_inserted = mod_score_insertions_removed\n",
    "\n",
    "                elif len(mod_score_insertions_removed) < max(sequence_deletions):\n",
    "\n",
    "                    mod_score_deletions_inserted_padded = np.pad(\n",
    "                        mod_score_insertions_removed, (0, \n",
    "                                                        sequence_length - len(mod_score_insertions_removed)),\n",
    "                                                    constant_values=0)\n",
    "\n",
    "\n",
    "                    mod_score_deletions_inserted = np.insert(mod_score_deletions_inserted_padded,\n",
    "                                                             sequence_deletions,\n",
    "                                                             0)\n",
    "                else: \n",
    "                    mod_score_deletions_inserted = np.insert(mod_score_insertions_removed,\n",
    "                                                             sequence_deletions,\n",
    "                                                             0)\n",
    "                mod_score = mod_score_deletions_inserted\n",
    "                \n",
    "                if (region_end_index > read_end_position) and (region_start_index > read_start_position): \n",
    "                    mod_start = 0\n",
    "                    mod_end = sequence_length\n",
    "                \n",
    "                # scenario 1: when the defined region is inside the read\n",
    "                elif (read_start_position <= region_start_index) and (read_end_position >= region_end_index):\n",
    "                    mod_start = region_start_index - read_start_position\n",
    "                    mod_end = region_end_index - read_start_position\n",
    "                \n",
    "                # scenario 3: when the defined region covers a bit of the end of the read\n",
    "                elif (read_end_position < region_end_index) and (read_end_position > region_start_index):\n",
    "                    mod_start = region_start_index - read_start_position \n",
    "                    mod_end = sequence_length\n",
    "                \n",
    "                # scenario 2: when the defined region covers a bit of the beginning of the read\n",
    "                elif (read_start_position > region_start_index) and (read_start_position < region_end_index):\n",
    "                    mod_start = 0\n",
    "                    mod_end = region_end_index - read_start_position \n",
    "                \n",
    "                #use the defined starting and ending positons in the region to subset mod numpy\n",
    "                trimmed_mod_score = mod_score[mod_start:mod_end]\n",
    "                \n",
    "                #removing all the zeros \n",
    "                mod_no_zeros = trimmed_mod_score[trimmed_mod_score != 0]\n",
    "                mA = len (mod_no_zeros)\n",
    "                \n",
    "                #Getting the total amount of As in the subsetted region of the sequence \n",
    "                total_A = true_sequence[mod_start:mod_end].count(\"A\")\n",
    "                \n",
    "                \n",
    "                #calculate read density\n",
    "                try:\n",
    "                    read_density = mA / total_A\n",
    "                    \n",
    "                except ZeroDivisionError:\n",
    "                    pass\n",
    "                region_density.append (read_density)\n",
    "\n",
    "                \n",
    "                    \n",
    "                \n",
    "            #calculate averaged region density average \n",
    "            try:\n",
    "                region_density_average = sum(region_density)/len(region_density)\n",
    "            \n",
    "            except ZeroDivisionError:\n",
    "                region_density_average = 0\n",
    "            data_table.append ([chr_name,region,region_density_average])\n",
    "\n",
    "        \n",
    "    table = tabulate(data_table, headers=\"firstrow\", tablefmt=\"fancy_grid\", floatfmt=\".18f\")\n",
    "    \n",
    "    filename = f\"{name}_region_density_scores.csv\"\n",
    "    with open(filename, \"w\", newline=\"\") as csvfile:\n",
    "        writer = csv.writer(csvfile,delimiter=\"\\t\")\n",
    "        writer.writerows(data_table)\n",
    "\n",
    "    print (name)\n",
    "    print (table)\n",
    "\n",
    "#region_read_mA_density_calculator(test_dict,'test_dict')\n",
    "\n",
    "#for scenario in scenarios: \n",
    "    #region_read_mA_density_calculator (scenario,str(scenario))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4651219a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 529,
   "id": "799e17d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrom_X_CDR_dict_Ms\n",
      "╒═════════════════╤════════════════════════════╤═════════════════════════╕\n",
      "│ chrX_MATERNAL   │ ['59282768', '59299767']   │   0.0005460337404030495 │\n",
      "╞═════════════════╪════════════════════════════╪═════════════════════════╡\n",
      "│ chrX_MATERNAL   │ ['59415768', '59448767']   │    0.001174757070153008 │\n",
      "├─────────────────┼────────────────────────────┼─────────────────────────┤\n",
      "│ chrX_MATERNAL   │ ['59461768', '59496767']   │    0.000886125639382954 │\n",
      "╘═════════════════╧════════════════════════════╧═════════════════════════╛\n",
      "chromosome_arm_random_region_dict_chrX_Ms\n",
      "╒═════════════════╤════════════════════════╤═════════════════════════╕\n",
      "│ chrX_MATERNAL   │ [37350684, 40463241]   │   0.0015864973155825344 │\n",
      "╞═════════════════╪════════════════════════╪═════════════════════════╡\n",
      "│ chrX_MATERNAL   │ [6225114, 9337671]     │    0.001385727181242197 │\n",
      "├─────────────────┼────────────────────────┼─────────────────────────┤\n",
      "│ chrX_MATERNAL   │ [143177622, 146290179] │    0.001228799034427388 │\n",
      "├─────────────────┼────────────────────────┼─────────────────────────┤\n",
      "│ chrX_MATERNAL   │ [68476254, 71588811]   │    0.001729275715637176 │\n",
      "╘═════════════════╧════════════════════════╧═════════════════════════╛\n",
      "chrom_X_CDR_adjacent_dict_Ms\n",
      "╒═════════════════╤════════════════════════╤══════════════════════╕\n",
      "│ chrX_MATERNAL   │ [59281767, 59282767]   │                  0.0 │\n",
      "╞═════════════════╪════════════════════════╪══════════════════════╡\n",
      "│ chrX_MATERNAL   │ [59299768, 59300768]   │ 0.000978436033948305 │\n",
      "├─────────────────┼────────────────────────┼──────────────────────┤\n",
      "│ chrX_MATERNAL   │ [59414767, 59415767]   │ 0.002198555389790337 │\n",
      "├─────────────────┼────────────────────────┼──────────────────────┤\n",
      "│ chrX_MATERNAL   │ [59448768, 59449768]   │ 0.001254208754208754 │\n",
      "├─────────────────┼────────────────────────┼──────────────────────┤\n",
      "│ chrX_MATERNAL   │ [59460767, 59461767]   │ 0.001119999005550272 │\n",
      "├─────────────────┼────────────────────────┼──────────────────────┤\n",
      "│ chrX_MATERNAL   │ [59496768, 59497768]   │ 0.001464493432445237 │\n",
      "╘═════════════════╧════════════════════════╧══════════════════════╛\n"
     ]
    }
   ],
   "source": [
    "from joblib import Parallel, delayed\n",
    "all_dicts = [chromosome_arm_random_region_dict, CDR_adjacent, CDR_dict]\n",
    "all_names = ['chromosome_arm_random_region_dict', 'CDR_adjacent', 'CDR_dict']\n",
    "def get_variable_name(var, locals_dict):\n",
    "    for name, value in locals_dict.items():\n",
    "        if value is var:\n",
    "            return name\n",
    "    return None\n",
    "\n",
    "\n",
    "\n",
    "#results=Parallel(n_jobs=3)(delayed(region_read_mA_density_calculator)(all_dicts,all_names) for i in range(3))\n",
    "\n",
    "for dictionary in range (0, len (chromX)): \n",
    "    variable_name = get_variable_name(chromX[dictionary], locals())\n",
    "\n",
    "    region_read_mA_density_calculator (chromX[dictionary],str(f\"{variable_name}_Ms\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d8fc269",
   "metadata": {},
   "source": [
    "'''version 1 '''\n",
    "''' \n",
    "The input of the function is a dictionary in the format of 'chromosome':[[start,end],[start,end]] \n",
    "'''\n",
    "def region_read_mA_density_calculator (chromosome_coordinates,name): \n",
    "    data_table = [] \n",
    "\n",
    "    #get each chromosome\n",
    "    for chr_name in chromosome_coordinates:\n",
    "\n",
    "        for region in chromosome_coordinates[chr_name]:\n",
    "            region_density = []\n",
    "\n",
    "            region_start_index = int(region[0])\n",
    "            region_end_index = int(region[1])\n",
    "            for read in bamfile.fetch(chr_name,region_start_index,region_end_index):\n",
    "                #checked read amoounts and names √\n",
    "                #make an if statement to check a specific read front, middle, end regions \n",
    "                if read.query_name == 'c06221ac-7a04-496c-87aa-e8f6bce21c34':\n",
    "\n",
    "                #setting read start, end, density, length variables \n",
    "                    \n",
    "                    #Get the starting and ending positions of the reads \n",
    "                    read_start_position = read.reference_start\n",
    "                    read_end_position = read.reference_end\n",
    "                    \n",
    "                    print ('start:', read_start_position)\n",
    "                    print ('end:', read_end_position)\n",
    "                    \n",
    "                    read_density = 0 \n",
    "                    \n",
    "                    \n",
    "                    sequence = read.get_aligned_pairs(matches_only=False, with_seq = True)\n",
    "                    count_none_first = sum(1 for item in sequence if item[1] is None)\n",
    "                    \n",
    "                    \n",
    "                    print ('sequence item:',count_none_first)\n",
    "                    sequence_length = len(sequence) - count_none_first\n",
    "                    print ('recalculated sequence length:', sequence_length)\n",
    "\n",
    "                    #removing reads shorter than 50000 \n",
    "                    if sequence_length < 50000:\n",
    "                        continue \n",
    "\n",
    "                    #make a mod np array with the length of the read length\n",
    "                    mod=read.modified_bases_forward\n",
    "                    mod_score = np.zeros(sequence_length,)\n",
    "\n",
    "\n",
    "                    #make transfer mA positions to mod np array corresponded to their sequence positions \n",
    "                    try:\n",
    "                        for indices, values in mod[('A', 0, 'a')]:\n",
    "                            mod_score[indices] = values\n",
    "\n",
    "                    except KeyError:\n",
    "                        continue \n",
    "\n",
    "                    #last thing to check is the region cutter \n",
    "                    if (read_start_position <= region_start_index) and (read_end_position >= region_end_index):\n",
    "                        print (\"scenario 1: when the defined region is inside the read\")\n",
    "                        mod_start = region_start_index - read_start_position\n",
    "                        mod_end = region_end_index - read_start_position\n",
    "                        \n",
    "                    elif (read_end_position < region_end_index) and (read_end_position > region_start_index):\n",
    "                        print (\"scenario 3: when the defined region covers a bit of the end of the read\")\n",
    "                        mod_start = region_start_index - read_start_position \n",
    "                        mod_end = sequence_length\n",
    "                    elif (read_start_position > region_start_index) and (read_start_position < region_end_index):\n",
    "                        print (\"scenario 2: when the defined region covers a bit of the beginning of the read\")\n",
    "                        mod_start = 0\n",
    "                        mod_end = region_end_index - read_start_position \n",
    "                    else: \n",
    "                        mod_start = 0\n",
    "                        mod_end = sequence_length\n",
    "                    \n",
    "                    \n",
    "                    print (mod_start,mod_end)\n",
    "                    trimmed_mod_score = mod_score[mod_start:mod_end]\n",
    "                    trimmed_sequence = read.query_sequence[mod_start:mod_end]\n",
    "                    print (trimmed_sequence)\n",
    "                    mod_threshold = 200\n",
    "                    mask = trimmed_mod_score <= mod_threshold\n",
    "                    trimmed_mod_score[mask] = 0 \n",
    "                    mod_no_zeros = trimmed_mod_score[trimmed_mod_score != 0]\n",
    "                    mA = len(mod_no_zeros)\n",
    "                    #checked filtered mA amount when reads are not sectioned into regions √\n",
    "\n",
    "                    # should we do total A in the assembly or use the total A amount ? \n",
    "\n",
    "                    #total_A = assembly[chr_name][(mod_start+read_start_position):(mod_end+read_start_position)].count(\"A\")\n",
    "                    total_A = read.query_sequence[mod_start:mod_end].count(\"A\")\n",
    "                    try:\n",
    "                        read_density = mA / total_A\n",
    "                    except ZeroDivisionError:\n",
    "                        read_density = 0\n",
    "                    region_density.append (read_density)\n",
    "\n",
    "                    print (region_density)\n",
    "                    \n",
    "\n",
    "                \n",
    "\n",
    "            try:\n",
    "                region_density_average = sum(region_density)/len(region_density)\n",
    "            except ZeroDivisionError:\n",
    "                region_density_average = 0\n",
    "            data_table.append ([chr_name,region,region_density_average])\n",
    "            \n",
    "            \n",
    "    table = tabulate(data_table, headers=\"firstrow\", tablefmt=\"fancy_grid\", floatfmt=\".18f\")\n",
    "\n",
    "    filename = f\"{name}_region_density_scores.csv\"\n",
    "    with open(filename, \"w\", newline=\"\") as csvfile:\n",
    "        writer = csv.writer(csvfile,delimiter=\"\\t\")\n",
    "        writer.writerows(data_table)\n",
    "\n",
    "    print (name)\n",
    "    print (table)\n",
    "\n",
    "#region_read_mA_density_calculator(test_dict,'test_dict')\n",
    "\n",
    "for scenario in scenarios: \n",
    "    region_read_mA_density_calculator (scenario,str(scenario))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "c72e0b4a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The name of the variable is: my_variable\n"
     ]
    }
   ],
   "source": [
    "def get_variable_name(var, locals_dict):\n",
    "    for name, value in locals_dict.items():\n",
    "        if value is var:\n",
    "            return name\n",
    "    return None\n",
    "\n",
    "# Example usage\n",
    "my_variable = 42\n",
    "variable_name = get_variable_name(my_variable, locals())\n",
    "print(f\"The name of the variable is: {variable_name}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "641cdd8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "''' \n",
    "version 2 \n",
    "'''\n",
    "def region_read_mA_density_calculator (chromosome_coordinates,name): \n",
    "    data_table = [] \n",
    "\n",
    "    #get each chromosome\n",
    "    for chr_name in chromosome_coordinates:\n",
    "\n",
    "        for region in chromosome_coordinates[chr_name]:\n",
    "            region_density = []\n",
    "\n",
    "            region_start_index = int(region[0])\n",
    "            region_end_index = int(region[1])\n",
    "            for read in bamfile.fetch(chr_name,region_start_index,region_end_index):\n",
    "                #checked read amoounts and names √\n",
    "                #make an if statement to check a specific read front, middle, end regions \n",
    "\n",
    "                #setting read start, end, density, length variables \n",
    "                    \n",
    "                #Get the starting and ending positions of the reads \n",
    "                read_start_position = read.reference_start\n",
    "                read_end_position = read.reference_end\n",
    "\n",
    "                read_density = 0 \n",
    "        \n",
    "                #Get sequence information which shows deletions and insertions \n",
    "                sequence = read.get_aligned_pairs(matches_only=False, with_seq = True)\n",
    "                \n",
    "                #take sequence length excluding insertions \n",
    "                insertion_amount = sum(1 for item in sequence if item[1] is None)\n",
    "                sequence_length = len(sequence) - insertion_amount\n",
    "                #sequence_length = len(sequence)\n",
    "\n",
    "                #removing reads shorter than 50000 \n",
    "                if sequence_length < 50000:\n",
    "                    continue \n",
    "\n",
    "                #make a mod np array with the length of the read length\n",
    "                mod=read.modified_bases_forward\n",
    "                mod_score = np.zeros(sequence_length,)\n",
    "                \n",
    "\n",
    "\n",
    "                #make transfer mA positions to mod np array corresponded to their sequence positions \n",
    "                try:\n",
    "                    for indices, values in mod[('A', 0, 'a')]:\n",
    "                        mod_score[indices] = values\n",
    "                \n",
    "                # No mod would return KeyError \n",
    "                except KeyError:\n",
    "                    continue \n",
    "                    \n",
    "                except IndexError:\n",
    "                    print (mod[('A', 0, 'a')])\n",
    "                    \n",
    "                    print (\"mA mod length:\",mod[('A', 0, 'a')][-1])\n",
    "                    print (\"mod_Score length:\",len(mod_score))\n",
    "                    print ('read_start_position', read_start_position)\n",
    "                    print ('read_end_position', read_end_position)\n",
    "                    print ('difference', read_end_position - read_start_position)\n",
    "                    \n",
    "\n",
    "                # scenario 1: when the defined region is inside the read\n",
    "                if (read_start_position <= region_start_index) and (read_end_position >= region_end_index):\n",
    "                    mod_start = region_start_index - read_start_position\n",
    "                    mod_end = region_end_index - read_start_position\n",
    "                \n",
    "                # scenario 3: when the defined region covers a bit of the end of the read\n",
    "                elif (read_end_position < region_end_index) and (read_end_position > region_start_index):\n",
    "                    mod_start = region_start_index - read_start_position \n",
    "                    mod_end = sequence_length\n",
    "                \n",
    "                # scenario 2: when the defined region covers a bit of the beginning of the read\n",
    "                elif (read_start_position > region_start_index) and (read_start_position < region_end_index):\n",
    "                    mod_start = 0\n",
    "                    mod_end = region_end_index - read_start_position \n",
    "                \n",
    "                # scenario 4: defined region covers the entire read \n",
    "                else: \n",
    "                    mod_start = 0\n",
    "                    mod_end = sequence_length\n",
    "\n",
    "                #use the defined starting and ending positons in the region to subset mod numpy\n",
    "                trimmed_mod_score = mod_score[mod_start:mod_end]\n",
    "                \n",
    "                #removing all the zeros \n",
    "                mod_no_zeros = trimmed_mod_score[trimmed_mod_score != 0]\n",
    "                \n",
    "                #Getting the total amount of As in the subsetted region of the sequence \n",
    "                total_A = read.query_sequence[mod_start:mod_end].count(\"A\")\n",
    "                \n",
    "                #calculate read density\n",
    "                try:\n",
    "                    read_density = mA / total_A\n",
    "                except ZeroDivisionError:\n",
    "                #if there is no A in the region, density becomes 0 \n",
    "                    read_density = 0\n",
    "                region_density.append (read_density)\n",
    "                \n",
    "\n",
    "            #calculate averaged region density average \n",
    "            try:\n",
    "                region_density_average = sum(region_density)/len(region_density)\n",
    "            \n",
    "            except ZeroDivisionError:\n",
    "                region_density_average = 0\n",
    "            data_table.append ([chr_name,region,region_density_average])\n",
    "            \n",
    "            \n",
    "    table = tabulate(data_table, headers=\"firstrow\", tablefmt=\"fancy_grid\", floatfmt=\".18f\")\n",
    "\n",
    "    filename = f\"{name}_region_density_scores.csv\"\n",
    "    with open(filename, \"w\", newline=\"\") as csvfile:\n",
    "        writer = csv.writer(csvfile,delimiter=\"\\t\")\n",
    "        writer.writerows(data_table)\n",
    "\n",
    "    print (name)\n",
    "    print (table)\n",
    "\n",
    "#region_read_mA_density_calculator(test_dict,'test_dict')\n",
    "\n",
    "#for scenario in scenarios: \n",
    "    #region_read_mA_density_calculator (scenario,str(scenario))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "id": "4de075da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0 11 1 2 3 4 12 5 6 7 13 8 9 nan nan 14 nan]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "# Create a NumPy array with None values\n",
    "my_array = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], dtype=object)\n",
    "\n",
    "# List of indices to add items\n",
    "indices_to_add = [1, 5, 8, 12]\n",
    "\n",
    "# Values to add at corresponding indices\n",
    "values_to_add = [11, 12, 13, 14]\n",
    "\n",
    "# Determine the desired length for the array\n",
    "desired_length = max(indices_to_add) + 1\n",
    "\n",
    "# Pad the array with np.nan values to the desired length\n",
    "padded_array = np.pad(my_array, (0, desired_length - len(my_array)), constant_values=np.nan)\n",
    "\n",
    "# Use NumPy's insert function to add items at specified indices\n",
    "result_array = np.insert(padded_array, indices_to_add, values_to_add)\n",
    "\n",
    "print(result_array)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a0c9764",
   "metadata": {},
   "outputs": [],
   "source": [
    "''' \n",
    "version 3 \n",
    "'''\n",
    "def region_read_mA_density_calculator (chromosome_coordinates,name): \n",
    "    data_table = [] \n",
    "\n",
    "    #get each chromosome\n",
    "    for chr_name in chromosome_coordinates:\n",
    "\n",
    "        for region in chromosome_coordinates[chr_name]:\n",
    "            region_density = []\n",
    "\n",
    "            region_start_index = int(region[0])\n",
    "            region_end_index = int(region[1])\n",
    "            for read in bamfile.fetch(chr_name,region_start_index,region_end_index):\n",
    "                #checked read amoounts and names √\n",
    "                #make an if statement to check a specific read front, middle, end regions \n",
    "\n",
    "                #setting read start, end, density, length variables \n",
    "                    \n",
    "                #Get the starting and ending positions of the reads \n",
    "                read_start_position = read.reference_start\n",
    "                read_end_position = read.reference_end\n",
    "\n",
    "                read_density = 0 \n",
    "        \n",
    "                #Get sequence information which shows deletions and insertions \n",
    "                sequence = read.get_aligned_pairs(matches_only=False, with_seq = True)\n",
    "                #make a numpy of the sequence length which eliminates the deletion\n",
    "                \n",
    "                print ('sequence length_get_aligned_pairs:', len(read.get_aligned_pairs(matches_only=False, with_seq = True)) - sum(1 for item in sequence if item[0] is None))\n",
    "                print ('sequence forward sequence:', len(read.query_sequence))\n",
    "                       \n",
    "                #take sequence length excluding insertions \n",
    "                insertion_amount = sum(1 for item in sequence if item[1] is None)\n",
    "                sequence_length = len(sequence) - insertion_amount\n",
    "                \n",
    "                \n",
    "                \n",
    "\n",
    "                #removing reads shorter than 50000 \n",
    "                if sequence_length < 50000:\n",
    "                    continue \n",
    "\n",
    "                #make a mod np array with the length of the read length\n",
    "                mod=read.modified_bases_forward\n",
    "                mod_score = np.zeros(sequence_length,)\n",
    "                \n",
    "                #make a mod score with its original length \n",
    "\n",
    "                #make transfer mA positions to mod np array corresponded to their sequence positions \n",
    "                try:\n",
    "                    for indices, values in mod[('A', 0, 'a')]:\n",
    "                        mod_score[indices] = values\n",
    "                \n",
    "                #eliminate the insertions in the mod score numpy \n",
    "                \n",
    "                # No mod would return KeyError \n",
    "                except KeyError:\n",
    "                    continue \n",
    "                    \n",
    "                except IndexError:\n",
    "                    print (mod[('A', 0, 'a')])\n",
    "                    sequence_length = len(read.get_aligned_pairs(matches_only=False, with_seq = True))\n",
    "                    print ('sequence length:', len(read.get_aligned_pairs(matches_only=False, with_seq = True)))\n",
    "                    print (\"mA mod length:\",mod[('A', 0, 'a')][-1])\n",
    "                    print (\"mod_Score length:\",len(mod_score))\n",
    "                    print ('read_start_position', read_start_position)\n",
    "                    print ('read_end_position', read_end_position)\n",
    "                    print ('difference', read_end_position - read_start_position)\n",
    "                    \n",
    "\n",
    "                # scenario 1: when the defined region is inside the read\n",
    "                if (read_start_position <= region_start_index) and (read_end_position >= region_end_index):\n",
    "                    mod_start = region_start_index - read_start_position\n",
    "                    mod_end = region_end_index - read_start_position\n",
    "                \n",
    "                # scenario 3: when the defined region covers a bit of the end of the read\n",
    "                elif (read_end_position < region_end_index) and (read_end_position > region_start_index):\n",
    "                    mod_start = region_start_index - read_start_position \n",
    "                    mod_end = sequence_length\n",
    "                \n",
    "                # scenario 2: when the defined region covers a bit of the beginning of the read\n",
    "                elif (read_start_position > region_start_index) and (read_start_position < region_end_index):\n",
    "                    mod_start = 0\n",
    "                    mod_end = region_end_index - read_start_position \n",
    "                \n",
    "                # scenario 4: defined region covers the entire read \n",
    "                else: \n",
    "                    mod_start = 0\n",
    "                    mod_end = sequence_length\n",
    "\n",
    "                #use the defined starting and ending positons in the region to subset mod numpy\n",
    "                trimmed_mod_score = mod_score[mod_start:mod_end]\n",
    "                \n",
    "                #removing all the zeros \n",
    "                mod_no_zeros = trimmed_mod_score[trimmed_mod_score != 0]\n",
    "                \n",
    "                #Getting the total amount of As in the subsetted region of the sequence \n",
    "                total_A = read.query_sequence[mod_start:mod_end].count(\"A\")\n",
    "                \n",
    "                #calculate read density\n",
    "                try:\n",
    "                    read_density = mA / total_A\n",
    "                except ZeroDivisionError:\n",
    "                #if there is no A in the region, density becomes 0 \n",
    "                    read_density = 0\n",
    "                region_density.append (read_density)\n",
    "                \n",
    "\n",
    "            #calculate averaged region density average \n",
    "            try:\n",
    "                region_density_average = sum(region_density)/len(region_density)\n",
    "            \n",
    "            except ZeroDivisionError:\n",
    "                region_density_average = 0\n",
    "            data_table.append ([chr_name,region,region_density_average])\n",
    "            \n",
    "            \n",
    "    table = tabulate(data_table, headers=\"firstrow\", tablefmt=\"fancy_grid\", floatfmt=\".18f\")\n",
    "\n",
    "    filename = f\"{name}_region_density_scores.csv\"\n",
    "    with open(filename, \"w\", newline=\"\") as csvfile:\n",
    "        writer = csv.writer(csvfile,delimiter=\"\\t\")\n",
    "        writer.writerows(data_table)\n",
    "\n",
    "    print (name)\n",
    "    print (table)\n",
    "\n",
    "#region_read_mA_density_calculator(test_dict,'test_dict')\n",
    "\n",
    "#for scenario in scenarios: \n",
    "    #region_read_mA_density_calculator (scenario,str(scenario))\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
